### PR TITLE
hal/ia32: Refactor timers

### PR DESCRIPTION
**Pull request by astalke**
## Description

Make timers more modular and add displaying description string about them.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: ia32-generic-qemu, ia32-generic-pc

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: hal/armv7m/stm32/l4/timer.c [hunk 1]
```diff
@@ -228,6 +229,13 @@ int hal_timerRegister(int (*f)(unsigned int, cpu_context_t *, void *), void *dat
 }
 
 
+char *hal_timerFeatures(char *features, unsigned int len)
+{
+	hal_strncpy(features, "Using default timer", len);
```
Comment (gerard5) [inline hal/armv7m/stm32/l4/timer.c:234]: STM32L4 use LPTIM, maybe:
```C
	hal_strncpy(features, "Using Low-Power Timer", len);
```

File: hal/armv7m/stm32/l1/timer.c [hunk 2]
```diff
@@ -48,6 +49,13 @@ void timer_jiffiesAdd(time_t t)
 }
 
 
+char *hal_timerFeatures(char *features, unsigned int len)
+{
+	hal_strncpy(features, "Using default timer", len);
```
Comment (gerard5) [inline hal/armv7m/stm32/l1/timer.c:54]: On STM32L1 Maybe:
```C
	hal_strncpy(features, "Using SysTick timer", len);
```

File: hal/armv7a/zynq7000/timer.c [hunk 3]
```diff
@@ -134,6 +135,13 @@ static void hal_timerSetPrescaler(u32 freq)
 }
 
 
+char *hal_timerFeatures(char *features, unsigned int len)
+{
+	hal_strncpy(features, "Using default timer", len);
```
Comment (gerard5) [inline hal/armv7a/zynq7000/timer.c:140]: Zynq7k uses TTC
```C
	hal_strncpy(features, "Using Triple Timer Counter", len);
```

File: hal/armv7a/imx6ull/timer.c [hunk 4]
```diff
@@ -123,6 +124,13 @@ int hal_timerRegister(int (*f)(unsigned int, cpu_context_t *, void *), void *dat
 }
 
 
+char *hal_timerFeatures(char *features, unsigned int len)
+{
+	hal_strncpy(features, "Using default timer", len);
```
Comment (gerard5) [inline hal/armv7a/imx6ull/timer.c:129]: On 6ULL two timers EPIT and GPT involved.

**Review by gerard5 — COMMENTED** [review]
Maybe instead of a name like "default timer", let's name them what they are actually used, what do you think?

File: hal/ia32/timer.c [hunk 5]
```diff
@@ -368,63 +479,157 @@ time_t hal_timerGetUs(void)
...
 	}
-	hal_spinlockClear(&timer.sp, &sc);
+	hal_spinlockClear(&timer_common.sp, &sc);
 }
 
 
+int hal_timerRegister(int (*f)(unsigned int, cpu_context_t *, void *), void *data, intr_handler_t *h)
+{
+	h->f = f;
+	h->n = SYSTICK_IRQ;
```
Comment (agkaminski) [inline hal/ia32/timer.c:508]: Should be `timer_common.schedulerTimer->schedulerIrq` instead of `SYSTICK_IRQ`

File: hal/ia32/timer.c [hunk 6]
```diff
@@ -49,56 +50,95 @@
...
 /* Programmable Interval Timer (Intel 8253/8254) */
 
+static unsigned int _hal_pitName(char *s, unsigned int *len)
+{
+	static const char prefix[] = "Programmable Interval Timer (";
+	static const char suffix[] = "kHz)";
+	unsigned int off = 0, n;
+	hal_strncpy(s + off, prefix, *len);
+	off += sizeof(prefix) - 1;
+	*len -= sizeof(prefix) - 1;
```
Comment (agkaminski) [inline hal/ia32/timer.c:124]: Applies to all `*name` functions: `len` might go negative here

File: hal/ia32/timer.c [hunk 7]
```diff
@@ -49,56 +50,95 @@
...
+	n = _hal_sprintU32(s + off, *len, PIT_FREQUENCY);
+	off += n;
+	*len -= n;
```
Comment (agkaminski) [inline hal/ia32/timer.c:127]: Applies to all `*name` functions: `len` might go negative here

File: hal/ia32/timer.c [hunk 8]
```diff
@@ -49,56 +50,95 @@
...
+	hal_strncpy(s + off, suffix, *len);
+	off += sizeof(suffix) - 1;
+	*len -= sizeof(suffix) - 1;
```
Comment (agkaminski) [inline hal/ia32/timer.c:130]: Applies to all `*name` functions: `len` might go negative here

File: hal/ia32/timer.c [hunk 9]
```diff
@@ -368,63 +479,157 @@ time_t hal_timerGetUs(void)
...
+	unsigned int off = 0;
+	(void)hal_strncpy(features + off, textScheduling, len);
+	off += textSchedulingLength;
+	len -= textSchedulingLength;
+
+	off += timer_common.schedulerTimer->name(features + off, &len);
+
+	(void)hal_strncpy(features + off, textTimestamp, len);
+	off += textTimestampLength;
+	len -= textTimestampLength;
```
Comment (agkaminski) [inline hal/ia32/timer.c:531]: `len` might go negative here

File: hal/ia32/timer.c [hunk 10]
```diff
@@ -368,63 +479,157 @@ time_t hal_timerGetUs(void)
...
+	.timestampGetUs = _hal_hpetGetUs,
+	.timestampBusyWaitUs = _hal_hpetBusyWaitUs,
+};
+
+
+/* List of scheduling timers from the best, to the worst */
+static const hal_timer_t *const _hal_schedulingTimers[] = { &_hal_lapicTimer };
+
+/* List of timestamp timers from the best, to the worst */
+static const hal_timer_t *const _hal_timestampTimers[] = { &_hal_hpetTimer };
```
Comment (agkaminski) [inline hal/ia32/timer.c:592]: Hmm, I don't think we expect more timers anytime soon, so no need for one-element lists for now

File: hal/armv7a/imx6ull/timer.c [hunk 11]
```diff
@@ -123,6 +124,13 @@ int hal_timerRegister(int (*f)(unsigned int, cpu_context_t *, void *), void *dat
 }
 
 
+char *hal_timerFeatures(char *features, unsigned int len)
+{
+	hal_strncpy(features, "Using EPIT and GPT timers", len);
```
Comment (kemonats) [inline hal/armv7a/imx6ull/timer.c:129]: For safety:
```suggestion
	hal_strncpy(features, "Using EPIT and GPT timers", len);
	features[len - 1] = '\0';
```
also in other functions.

File: hal/ia32/timer.c [hunk 12]
```diff
@@ -49,56 +50,95 @@
...
+
+	struct {
+		hal_gasMapped_t addr;
+		u32 period;
+		intr_handler_t tmr1;
+	} hpetData;
+} timer_common;
+
+
+static unsigned int _hal_sprintU32(char *s, unsigned char len, u32 n)
```
Comment (kemonats) [inline hal/ia32/timer.c:95]: One more function to convert `integer` to `string`.
Why not `hal_i2s()`?

Comment (agkaminski) [reply hal/ia32/timer.c:95]: Right, it's even exported to hal API now too

**Review by agkaminski — APPROVED** [review]
LGTM

**Comment by kemonats** [comment]
This PR has been merged but I have a few comments.
In the function code (ia32) : `hal_timerFeatures(), _hal_hpetName(), _hal_lapicTimerName(), _hal_pitName()`
there is the possibility of writing outside the buffer range
Further some examples.

File: hal/ia32/timer.c [hunk 13]
```diff
@@ -49,56 +50,80 @@
...
+static unsigned int _hal_pitName(char *s, unsigned int *len)
+{
+	static const char suffix[] = "kHz)";
+	unsigned int off = 0, n;
 
-static int hal_pitTimerIrqHandler(unsigned int n, cpu_context_t *ctx, void *arg)
+	n = hal_i2s("Programmable Interval Timer (", s + off, PIT_FREQUENCY, 10, 0);
+	if (*len < n) {
+		n = *len;
+	}
```
Comment (kemonats) [inline hal/ia32/timer.c:105]: If `*len < n` function `hal_i2s()` writes out of range `s`.

File: hal/ia32/timer.c [hunk 14]
```diff
@@ -135,25 +160,83 @@ static inline u16 _hal_pitReadTimer(void)
...
 
+static unsigned int _hal_lapicTimerName(char *s, unsigned int *len)
+{
+	static const char suffix[] = "kHz)";
+	unsigned int off = 0, n;
+
+	n = hal_i2s("Local APIC Timer (", s + off, timer_common.lapicData.frequency, 10, 0);
+	if (*len < n) {
+		n = *len;
+	}
```
Comment (kemonats) [inline hal/ia32/timer.c:225]: Same as in function `_hal_pitName()`.

File: hal/ia32/timer.c [hunk 15]
```diff
@@ -197,45 +280,66 @@ static inline void _hal_lapicTimerConfigure(u32 mode, u32 mask, u32 vector)
...
 
+static unsigned int _hal_hpetName(char *s, unsigned int *len)
+{
+	static const char suffix[] = "kHz)";
+	unsigned int off = 0, n;
+
+	n = hal_i2s("High Precision Timer (", s + off, (u32)(1000000000000LLU / ((u64)timer_common.hpetData.period)), 10, 0);
+	if (*len < n) {
+		n = *len;
+	}
```
Comment (kemonats) [inline hal/ia32/timer.c:319]: Same as in function `_hal_pitName()`.

File: hal/ia32/timer.c [hunk 16]
```diff
@@ -368,63 +472,144 @@ time_t hal_timerGetUs(void)
...
+{
+	static const char textScheduling[] = "Scheduling timer = ";
+	static const char textTimestamp[] = "\nhal: Timestamp timer = ";
+	const size_t textSchedulingLength = sizeof(textScheduling) - 1;
+	const size_t textTimestampLength = sizeof(textTimestamp) - 1;
+
+	unsigned int off = 0;
+	(void)hal_strncpy(features + off, textScheduling, len);
+	off += textSchedulingLength;
+	len -= textSchedulingLength;
```
Comment (kemonats) [inline hal/ia32/timer.c:518]: What if `len < textSchedulingLength`

**Pull request merged**